### PR TITLE
955 add indexes

### LINKED ISSUE #955: Indexes
Would be good to index at least the FKs. Ideally we would wait until we observer slowness and do targeted indexing, but I think it's ok to index FKs and any other visible/probably performance issues. Can refer to https://github.com/openmsupply/conforma-server/issues/899 for some sql that will find all unindexed fk constraints

### PR DESCRIPTION
**Pull request by andreievg**
closes #955 

Used this to get not indexed FK's

<details>
<summary>
Getting FK indexes
</summary>

```sql
-- check for FKs where there is no matching index
-- on the referencing side
-- or a bad index

WITH fk_actions ( code, action ) AS (
    VALUES ( 'a', 'error' ),
        ( 'r', 'restrict' ),
        ( 'c', 'cascade' ),
        ( 'n', 'set null' ),
        ( 'd', 'set default' )
),
fk_list AS (
    SELECT pg_constraint.oid as fkoid, conrelid, confrelid as parentid,
        conname, relname, nspname,
        fk_actions_update.action as update_action,
        fk_actions_delete.action as delete_action,
        conkey as key_cols
    FROM pg_constraint
        JOIN pg_class ON conrelid = pg_class.oid
        JOIN pg_namespace ON pg_class.relnamespace = pg_namespace.oid
        JOIN fk_actions AS fk_actions_update ON confupdtype = fk_actions_update.code
        JOIN fk_actions AS fk_actions_delete ON confdeltype = fk_actions_delete.code
    WHERE contype = 'f'
),
fk_attributes AS (
    SELECT fkoid, conrelid, attname, attnum
    FROM fk_list
        JOIN pg_attribute
            ON conrelid = attrelid
            AND attnum = ANY( key_cols )
    ORDER BY fkoid, attnum
),
fk_cols_list AS (
    SELECT fkoid, array_agg(attname) as cols_list
    FROM fk_attributes
    GROUP BY fkoid
),
index_list AS (
    SELECT indexrelid as indexid,
        pg_class.relname as indexname,
        indrelid,
        indkey,
        indpred is not null as has_predicate,
        pg_get_indexdef(indexrelid) as indexdef
    FROM pg_index
        JOIN pg_class ON indexrelid = pg_class.oid
    WHERE indisvalid
),
fk_index_match AS (
    SELECT fk_list.*,
        indexid,
        indexname,
        indkey::int[] as indexatts,
        has_predicate,
        indexdef,
        array_length(key_cols, 1) as fk_colcount,
        array_length(indkey,1) as index_colcount,
        round(pg_relation_size(conrelid)/(1024^2)::numeric) as table_mb,
        cols_list
    FROM fk_list
        JOIN fk_cols_list USING (fkoid)
        LEFT OUTER JOIN index_list
            ON conrelid = indrelid
            AND (indkey::int2[])[0:(array_length(key_cols,1) -1)] @> key_cols

),
fk_perfect_match AS (
    SELECT fkoid
    FROM fk_index_match
    WHERE (index_colcount - 1) <= fk_colcount
        AND NOT has_predicate
        AND indexdef LIKE '%USING btree%'
),
fk_index_check AS (
    SELECT 'no index' as issue, *, 1 as issue_sort
    FROM fk_index_match
    WHERE indexid IS NULL
    UNION ALL
    SELECT 'questionable index' as issue, *, 2
    FROM fk_index_match
    WHERE indexid IS NOT NULL
        AND fkoid NOT IN (
            SELECT fkoid
            FROM fk_perfect_match)
),
parent_table_stats AS (
    SELECT fkoid, tabstats.relname as parent_name,
        (n_tup_ins + n_tup_upd + n_tup_del + n_tup_hot_upd) as parent_writes,
        round(pg_relation_size(parentid)/(1024^2)::numeric) as parent_mb
    FROM pg_stat_user_tables AS tabstats
        JOIN fk_list
            ON relid = parentid
),
fk_table_stats AS (
    SELECT fkoid,
        (n_tup_ins + n_tup_upd + n_tup_del + n_tup_hot_upd) as writes,
        seq_scan as table_scans
    FROM pg_stat_user_tables AS tabstats
        JOIN fk_list
            ON relid = conrelid
)
-- SELECT nspname as schema_name,
--     relname as table_name,
--     conname as fk_name,
--     issue,
--     table_mb,
--     writes,
--     table_scans,
--     parent_name,
--     parent_mb,
--     parent_writes,
--     cols_list,
--     indexdef
-- FROM fk_index_check
--     JOIN parent_table_stats USING (fkoid)
--     JOIN fk_table_stats USING (fkoid)
-- ORDER BY issue_sort, table_mb DESC, table_name, fk_name;

SELECT 'CREATE INDEX "i_' || conname || '" ON "' || relname || '" ("' || array_to_string(cols_list, '","') || '");' 
FROM fk_index_check
    JOIN parent_table_stats USING (fkoid)
    JOIN fk_table_stats USING (fkoid)
```
</details>

Then went through every table, and indexed everything that I thought needs an index.

### CODE REVIEW

File: server/repository/migrations/postgres/2021-09-18T10-00_create_requisition_line table/up.sql [hunk 1]
```diff
@@ -3,7 +3,7 @@
 CREATE TABLE requisition_line (
     id TEXT NOT NULL PRIMARY KEY,
     requisition_id TEXT NOT NULL REFERENCES requisition (id),
-    item_id TEXT NOT NULL,
+    item_id TEXT NOT NULL REFERENCES item(id),
```
Comment (andreievg) [inline server/repository/migrations/postgres/2021-09-18T10-00_create_requisition_line table/up.sql:6]: I thought I could add this with migration, but not easy with sqlite

File: server/repository/src/migrations/v1_01_01/mod.rs [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+            CREATE INDEX "index_item_unit_id_fkey" ON "item" ("unit_id");
+            CREATE INDEX "index_name_store_join_store_id_fkey" ON "name_store_join" ("store_id");
+            CREATE INDEX "index_name_store_join_name_id_fkey" ON "name_store_join" ("name_id");
+            CREATE INDEX "index_stock_line_location_id_fkey" ON "stock_line" ("location_id");
+            CREATE INDEX "index_stock_line_store_id_fkey" ON "stock_line" ("store_id");
+            CREATE INDEX "index_stock_line_item_id_fkey" ON "stock_line" ("item_id");
+            CREATE INDEX "index_master_list_line_master_list_id_fkey" ON "master_list_line" ("master_list_id");
+            CREATE INDEX "index_master_list_line_item_id_fkey" ON "master_list_line" ("item_id");
+            CREATE INDEX "index_requisition_line_requisition_id_fkey" ON "requisition_line" ("requisition_id");
+            CREATE INDEX "index_requisition_name_id_fkey" ON "requisition" ("name_id");
```
Comment (mark-prins) [inline server/repository/src/migrations/v1_01_01/mod.rs:53]: perhaps should move down, under `-- Requisition`

File: server/repository/src/migrations/v1_01_01/mod.rs [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+            CREATE INDEX "index_master_list_name_join_name_id_fkey" ON "master_list_name_join" ("name_id");
+            CREATE INDEX "index_item_unit_id_fkey" ON "item" ("unit_id");
+            CREATE INDEX "index_name_store_join_store_id_fkey" ON "name_store_join" ("store_id");
+            CREATE INDEX "index_name_store_join_name_id_fkey" ON "name_store_join" ("name_id");
+            CREATE INDEX "index_stock_line_location_id_fkey" ON "stock_line" ("location_id");
+            CREATE INDEX "index_stock_line_store_id_fkey" ON "stock_line" ("store_id");
+            CREATE INDEX "index_stock_line_item_id_fkey" ON "stock_line" ("item_id");
+            CREATE INDEX "index_master_list_line_master_list_id_fkey" ON "master_list_line" ("master_list_id");
+            CREATE INDEX "index_master_list_line_item_id_fkey" ON "master_list_line" ("item_id");
+            CREATE INDEX "index_requisition_line_requisition_id_fkey" ON "requisition_line" ("requisition_id");
```
Comment (mark-prins) [inline server/repository/src/migrations/v1_01_01/mod.rs:52]: perhaps should move down, under `-- Requisition Line`

File: server/repository/src/migrations/v1_01_01/mod.rs [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+            -- Foreign Keys (defined as REFERENCES)
+            CREATE INDEX "index_stocktake_line_item_id_fkey" ON "stocktake_line" ("item_id");
+            CREATE INDEX "index_stocktake_line_location_id_fkey" ON "stocktake_line" ("location_id");
+            CREATE INDEX "index_stocktake_line_stock_line_id_fkey" ON "stocktake_line" ("stock_line_id");
+            CREATE INDEX "index_stocktake_line_stocktake_id_fkey" ON "stocktake_line" ("stocktake_id");
+            CREATE INDEX "index_invoice_line_location_id_fkey" ON "invoice_line" ("location_id");
+            CREATE INDEX "index_invoice_line_stock_line_id_fkey" ON "invoice_line" ("stock_line_id");
+            CREATE INDEX "index_invoice_line_item_id_fkey" ON "invoice_line" ("item_id");
+            CREATE INDEX "index_invoice_line_invoice_id_fkey" ON "invoice_line" ("invoice_id");
+            CREATE INDEX "index_activity_log_store_id_fkey" ON "activity_log" ("store_id");
```
Comment (mark-prins) [inline server/repository/src/migrations/v1_01_01/mod.rs:27]: not `record_id` as well? the usual lookup for the log is by the related record id

**Review by mark-prins — APPROVED** [review]
👍 LGTM!

One question - nothing for the `[name]` table?
is that not needed because we usually are referencing name_store_join which limits the results?

File: server/repository/src/migrations/v1_01_01/mod.rs (hunk 2 above)
Comment (andreievg) [reply server/repository/src/migrations/v1_01_01/mod.rs:53]: Could do, the first section combines all of the tables where index does not exist for references (relation), it was auto generated with a script in the comment, so they are just grouped by relation/reference.

Just a general note, with migrations, we wouldn't really be able to make sense of what the schema is by looking at migrations themselves. Could only really see the real state of DB through some tool (pgAdmin etc..), so I was thinking that migration files won't really be used to make sense of schema itself thus don't really need to be exceptionally tidy. There is a carry over in one of the issue for schema spy (for quick and easy way to see state of schema, including indexes)

File: server/repository/src/migrations/v1_01_01/mod.rs (hunk 3 above)
Comment (andreievg) [reply server/repository/src/migrations/v1_01_01/mod.rs:52]: Similar to comment above

File: server/repository/src/migrations/v1_01_01/mod.rs (hunk 4 above)
Comment (andreievg) [reply server/repository/src/migrations/v1_01_01/mod.rs:27]: 🙏 f21937a1c98d3c62f7a771c86f32da0bdb1a2eec

**Comment by andreievg** [comment]
@mark-prins 

> One question - nothing for the [name] table?

We would definitley need to do something for name as patient. The thinking was that in terms of facilities, name list would be in the thousands and most queries should be quick, without indexes. 

I'll try to mention this in standup, to make sure index strategy for patient names is not missed.

**Comment by andreievg** [comment]
🙏 for review

**Pull request merged**